File: Haskell/src/Statistics.hs
module Statistics where
    import Constants
    import GeneralPrints
    import Validation
    import System.IO.Unsafe
    import System.Directory
    import System.FilePath
    import Data.List
    import Control.Monad
    import Control.DeepSeq
    import Prelude hiding (readFile)
    import System.IO.Strict (readFile)

    import TestSuite
    import TestCase

    sumStatisticsProject :: Int -> [Suite] -> Float
    sumStatisticsProject _ [] = 0.0
    sumStatisticsProject projectId ((Suite id _ _ _):list) = (calculateStatiscs projectId id) + (sumStatisticsProject projectId list)

    calculateMediumStatisticsProject :: Int -> [Suite] -> Float
    calculateMediumStatisticsProject projectId suites = ((sumStatisticsProject projectId suites) / (fromIntegral (length suites)))

    getProjectResume :: [(Int, String)] -> String
    getProjectResume [] = []
    getProjectResume (((id, name)):projectsTouple) = (show id) ++ " - " ++ name ++ " - " ++ (show (calculateMediumStatisticsProject id (getSuitesList id))) ++ "%\n" ++ (getProjectResume projectsTouple)

    getProjectsStatistics :: [(Int, String)] -> IO()
    getProjectsStatistics projectsTouple = do
        putStrLn statistics_header
        putStrLn " ID - NOME DO PROJETO - MÉDIA DAS ESTATÍSTICAS DO PROJETO\n"
        putStrLn (getProjectResume projectsTouple)
        return ()


    getProjectsId :: [(Int, String)] -> [Int]
    getProjectsId [] = []
    getProjectsId ((id, name):list) = id:(getProjectsId list)

    chooseStatisticsAction :: Char -> [(Int, String)]  -> IO()
    chooseStatisticsAction option projectsTouple
        | option == '1' = do getProjectsStatistics projectsTouple
        | option == '2' = do statisticsFromAProject (getProjectsId projectsTouple)
        | otherwise = do print invalid_option


    isValidProjId :: Int -> [Int] -> Bool
    isValidProjId _ [] = False
    isValidProjId testedId (id:projectsId)
        | testedId == id = True
        | otherwise = isValidProjId testedId projectsId

    statisticsFromAProject :: [Int] -> IO()
    statisticsFromAProject projectsId = do
        putStrLn statistics_header
        putStrLn "Informe o ID de um projeto para visualizar seu relatório:"
        projectId <- getLine
        if isStringNumeric projectId
            then do
                if isValidProjId (read projectId) projectsId
                    then do
                        clearScreen
                        let suites = getSuitesList (read projectId)
                            outputString = generateStatisticsString (read projectId) suites
                        putStrLn statistics_header
                        putStrLn ("SUITE ID - NOME DA SUITE - TAXA DE TESTES QUE PASSARAM\n" ++ outputString)
                    else do
                        putStrLn ("Não existe projeto com o ID informado.")
            else do
                putStrLn "O ID informado é inválido."
    
    generateStatisticsString :: Int -> [Suite] -> String
    generateStatisticsString _ [] = []
    generateStatisticsString projectId ((Suite id name _ _):suites) = show id ++ " - " ++ name ++ " - " ++ (show (calculateStatiscs projectId id)) ++ "%\n" ++ (generateStatisticsString projectId suites)
    
    getSuitesList :: Int -> [Suite]
    getSuitesList projId = unsafePerformIO $ readSuites projId

    showStatisticsMenu :: IO()
    showStatisticsMenu = do
        printHeaderWithSubtitle statistics_menu 

    statisticsMenu :: [(Int, String)] -> IO()
    statisticsMenu projectsTouple = do
        showStatisticsMenu
        putStrLn choose_option
        input <- getLine
        if isOptionValid input '1' '3'
                then do
                    let option = input !! 0
                    if option == '3'
                        then putStrLn "Retornando ao menu anterior..."
                        else do
                            clearScreen
                            chooseStatisticsAction option projectsTouple
                            systemPause
                            statisticsMenu projectsTouple
                else do
                    putStrLn invalid_option
                    systemPause
                    statisticsMenu projectsTouple

File: Prolog/src/Statistics.pl
:- initialization(main).

:- use_module(util).

sumStatisticsProject(_,[], Result):- Result is 0.
sumStatisticsProject(ProjectId,[Id|List], Result) :-  sumStatisticsProject(ProjectId,List, Result2), 
                    testCase:calculateStatiscs(ProjectId, Id, StatSuite), Result is Result2 + StatSuite.

calculateMediumStatisticsProject(ProjectId, [H|T],Result) :- sumStatisticsProject(ProjectId, [H|T],Z), Y is util:length_1(Z,[H|T]), Result is Z/Y.


getProjectResume([]).
getProjectResume([[Id, Name]|ProjectsTouple]):- write(Id), write(" - "), write(Name), write(" - "), 
                    Suites is getSuitesList(Id),    
                    Media is calculateMediumStatisticsProject(Id,Suites),
                    write(Media),nl, 
                    getProjectResume(ProjectsTouple).

getProjectsStatistics([ProjectsTouple]) :-
    write("#--------------# ESTATÍSTICAS #------------#"),nl,
    write(" ID - NOME DO PROJETO - MÉDIA DAS ESTATÍSTICAS DO PROJETO"),nl,
    getProjectResume(ProjectsTouple).


getProjectId([], []).
getProjectId([(Id,Name)|List]) :- [Id|getProjectId(List)].

chooseStatisticsAction(Option, [(X,Y)]) :-
    (Option =:= "1" -> getProjectsStatistics([(X,Y)]);
    Option =:= "2" -> statisticsFromAProject(getProjectId([(X,Y)]);
    write("Opção Inválida"))).
    


isValidProjId(_,[]).
isValidProjId(TestedId, [Id|ProjectsId]):-
    (TestedId =:= Id -> true;
    isValidProjId(TestedId, ProjectsId)).

statisticsFromAProject(projectsId):-
    write("#--------------# ESTATÍSTICAS #------------#"),
    write("Informe o ID de um projeto para visualizar seu relatório:"),
    read_line_to_codes(user_input, ProjectId),
    (validation:isStringNumeric(ProjectId) = true -> 
        (validation:isValidProjId(ProjectId,projectsId) = true -> 
        tty_clear,
                    Suites is getSuitesList(ProjectId),
                    Saida is generateStatisticsString(ProjectId, Suites),
                    write("#--------------# ESTATÍSTICAS #------------#"),nl,
                    write("SUITE ID - NOME DA SUITE - TAXA DE TESTES QUE PASSARAM"),nl,
                    write(Saida);
                    write("Não existe projeto com o ID informado."));
    write("O ID informado é inválido.")).

generateStatisticsString(_,[], []).
generateStatisticsString(ProjectId,[Id, Name,_,_|Suites]):- 
        write(Id), write(" - "), write(Name), write(" - "), 
        testCase:calculateStatiscs(ProjectId, Id, StatSuite),nl, 
        generateStatisticsString(ProjectId, Suites).


getSuitesList(projId) = 0.

showStatisticsMenu() :-
    generalPrints:printHeaderWithSubtitle(statistics_menu).

statisticsMenu([Id, Y]) :-
    showStatisticsMenu(),
    write("Informe a opção desejada: "),
    read_line_to_codes(user_input, Entrada),
    (validation:isOptionValid(Entrada, "1", "3") = true ->
            Option is Entrada, (Option =:= "3" -> write("Retornando ao menu anterior...");
            tty_clear, chooseStatisticsAction(Option, [Id,Y]), generalPrints:systemPause(), statisticsMenu([Id,Y]));                
    write("Opção Inválida"),
    generalPrints:systemPause(),
    statisticsMenu([Id,Y])).